src/dbnode/storage/index.go

```diff
@@ -716,77 +723,196 @@ func (i *nsIndex) Query(
+		if opts.Limit > 0 && results.Size() >= opts.Limit {
+			results.Lock()
+			exhaustive = false
+			results.Unlock()
+			break
+		}
+
+		execBlockQuery := func() {
+			currExhaustive, err := block.Query(query, opts, results)
+			if err != index.ErrUnableToQueryBlockClosed {
+				// NB(r): Because we query this block outside of the lock, it's possible this
+				// block may get closed if it slides out of retention, in that case
+				// those results are no longer considered valid and outside of retention
+				// regardless, so this is a non-issue
+				err = nil
+			}
+
+			results.Lock()
+			defer results.Unlock()
+
+			if err != nil {
+				multiErr = multiErr.Add(err)
+				return
+			}
+
+			if !currExhaustive {
+				// i.e. block had more data but we stopped early, we know
+				// we have hit the limit and don't need to query any more.
+				exhaustive = false
+				return
+			}
+		}
+
+		if timeoutApplied := timeout > 0; !timeoutApplied {
+			// No timeout, just wait blockingly for a worker
+			wg.Add(1)
+			workers.Go(func() {
+				execBlockQuery()
+				wg.Done()
+			})
+			continue
+		}
+
+		// Need to apply timeout
+		timedOut := false
+		now := i.nowFn()
+		if timeLeft := deadline.Sub(now); timeLeft > 0 {
+			wg.Add(1)
+			running := workers.GoWithTimeout(func() {
+				execBlockQuery()
+				wg.Done()
+			}, timeLeft)
+			timedOut = !running
+		} else {
+			timedOut = true
+		}
+
+		if timedOut {
+			// Exceeded our deadline waiting for this block's query to start
+			return index.QueryResults{},
+				fmt.Errorf("index query exceeded timeout: %s", timeout.String())
+		}
+	}
+
+	// Wait for queries to finish
+	wg.Wait()
+
+	// NB(r): Dont' need to hold lock for concurrent results since
+	// at this point all block queries must have finished.
+	// If we ever to early aborts we'll have to make sure that the results
+	// aren't modified by aborted goroutines we launched that still have
+	// access to the concurrent results set (which is difficult because
+	// the reader of the results will also need to hold lock instead of just
+	// being returned unsynchronized results from this method like we
+	// currently do).
+	if err := multiErr.FinalError(); err != nil {
+		return index.QueryResults{}, err
+	}
+
+	return index.QueryResults{
+		Exhaustive: exhaustive,
+		Results:    results.Results(),
+	}, nil
+}
+
+func (i *nsIndex) timeoutForQueryWithLock(
+	ctx context.Context,
+) time.Duration {
+	// TODO(r): Allow individual queries to specify timeouts using
+	// deadlines passed by the context
+	return i.state.runtimeOpts.defaultQueryTimeout
+}
+
+func (i *nsIndex) overriddenOptsForQueryWithLock(
+	opts index.QueryOptions,
+) index.QueryOptions {
+	// Override query response limit if needed
 	if i.state.runtimeOpts.maxQueryLimit > 0 && (opts.Limit == 0 ||
 		int64(opts.Limit) > i.state.runtimeOpts.maxQueryLimit) {
 		i.logger.Debugf("overriding query response limit, requested: %d, max-allowed: %d",
```

Comment:
Nit: I think queries are serializable now (at least the idx package has a `String` method on the query struct.